### DISCUSSION TITLE
Proposal - gathering stack state during exception handling

### DISCUSSION THREAD

**Discussion by darrenoakey** [dotnet/csharplang#3725] [Language Ideas]
I come from a world where _nothing_ is more important than knowing what happened if something goes wrong.  If the code is slow you work on it - if there's a bug, you fix it.  If you can't explain to the traders _why_ something happened - nice men in uniform take your badge away and walk you to the door.

so - ever since postsharp first came out, we've used postsharp or fody to add an exception handler to every single method that takes parameters.   this just catches the exception - adds a specially created object to the exception data which records any parameter to the method and the this pointer - and then rethrows the exception.      Then, in the exception toString - we replace  
      MyClass:Method()
with 
      MyClass:Method( "bob", 1234 )  on [this is whatever myclass.tostring spits out]
in the stack trace

If you code mostly functionally, as more people are moving towards, it means the stack trace tells you _exactly_ what's going on - which trade was involved, what values were being compared and so on - and usually someone can work out both what the bug is and how to fix it purely from just looking at the stack trace.  No need to read code, guess at, or try and reproduce anything

I'd love that to be built in to the exception handling system - or at least have option to turn it on - I reckon it can save any developer a half a day a week easy.   It has no real performance impact, as you don't do _anything_ unless an exception is thrown - and even then, you don't do any real work on the "toString" - it's just putting a lot of stuff into a dictionary on the way up.    

Some people may be we worried about recording and then logging sensitive details like a password, but it's like 0.001% of your code that deals with anything sensitive (or in my case 0.0%)  - and for that maybe there could just be an annotation like [Sensitive] - on either a method or parameter.

**Comment by HaloFour** [comment]
Sounds like a request for the [runtime repository](https://github.com/dotnet/runtime/)

**Comment by darrenoakey** [comment]
I don't think so - because I don't want something that ends up with something that can be interrogated at runtime - I actually want to change the way the compiler generates code - so it generates this information in the standard stack trace.  I guess the worry is that people have to change code to use it...whereas I'm hoping that if you set a flag on the compiler, every thing that currently logs out an exception now logs out an exception with useful information.

maybe it _is_ something better to put into the runtime, I'm not sure, but I'm _currently_ doing it by weaving in code at compile time.

**Comment by canton7** [comment]
I think this might have an adverse effect on the runtime. Currently a method's parameters, and even `this`, can be collected before the method returns. This change would force those to live for the entire duration of the method.

It would also affect inlining: IIRC a method containing a `throw` is never inlined, but now every method would contain a `throw`, so inlining would not happen at all. That's a huge potential perf hit.

**Comment by darrenoakey** [comment]
I see and agree with both problems (although you'd think, if this is in the compiler, it doesn't necessarily have to follow the "don't inline throws" thing) but bottom line, for me - _nothing_ is as important as getting this information - _any_ performance hit is worth it - it's not ideal, but it's an easy choice.    eg - we even logged all network calls out and response to the app as well as every keydown, key up and and mouse movement - and I remember a particular incident where I had to fly to London to sit with a team of lawyers and explain to them exactly what certain log messages meant and how to read them.   Preventing the program screwing up is nice and all, but it has diminishing returns...    knowing why the program screwed up when it inevitably does is just as, or even more valuable.

**Comment by canton7** [comment]
What's the motivation for making this a language option, rather than using an IL weaver? What issues do you hit when using an IL weaver?

**Comment by CyrusNajmabadi** [comment]
I'm not even certain I understand what the language request is here. It sounds perhaps like this is a compiler or runtime request

**Comment by darrenoakey** [comment]
> What's the motivation for making this a language option, rather than using an IL weaver? What issues do you hit when using an IL weaver?

It actually works fine with a weaver - however - I'm a consultant and go into a number of companies - and it's hard to sell using a weaver - would be much easier to argue if it were a language feature

**Comment by darrenoakey** [comment]
> I'm not even certain I understand what the language request is here. It sounds perhaps like this is a compiler or runtime request

the thing is - the way I currently implement it is by dicking with the compiled code.  While there might be some alternative way of doing it in the runtime or otherwise - I don't know what that is, or how it would work - so I guess I've asked for it as a language feature because that's the only way I would know how to build it...

**Comment by CyrusNajmabadi** [comment]
> so I guess I've asked for it as a language feature because that's the only way I would know how to build it...

Could you clarify what the actual language change you are proposing is?  i.e. today i write `void Foo() { Console.WriteLine(); }`.  With your changes to the langauge, what would i write instead, and what would it mean?

**Comment by darrenoakey** [comment]
yeah - I see what you are saying.  I guess it's not a change to the language... it's a change to the _compiler_ - or the way code is generated, but it doesn't in any way change the way code is written.  I dunno - is this a different forum then?

**Comment by 333fred** [comment]
I would say either the compiler or the runtime repos. Personally, I would prefer to see this as a runtime feature, because it would make it more complete. There's no way to guarantee that you got all the methods with a weaver or the compile-time equivalent, someone could just use Reflection.Emit to create code, or load assemblies from disk. The runtime could actually make guarantees.

**Comment by john-h-k** [comment]
C# doesn't generate a stack trace, that is what the runtime does

**Comment by john-h-k** [comment]
> (although you'd think, if this is in the compiler, it doesn't necessarily have to follow the "don't inline throws" thing) 

Inlining is a runtime level operation, and is up to the runtime to do. `throw`s prevent inlining (almost always) because they massively bloat assembly size. Inlining methods with throws often isn't profitable.

This would absolutely destroy performance too. Why isn't breaking on exception throws in a debugger good enough?

**Reply by darrenoakey** [reply]
debugger?    I came from a world of e-trading.   Errors weren't meant to happen - however - if an error did... it wasn't so bad if you could explain it - and guarantee it would never happen again.   Literally nothing (even performance, although in a trading world, performance was important) - was as important as that.     to be able to explain what happened.  and now - when the trader has run over to and is standing behind you and asking.   You don't have half an hour to go and try to guess at the scenario, then try and duplicate the circumstances.  You need to _know_ what happened.

So - we used an IL weaver - every single method was surrounded by a try catch - so

```c#
int someFunction( string y, int z)
{
     return $"{y} - {z}";
}
```

actually turned into 

```c#
int someFunction( string y, int z)
{
    try
    {
     return $"{y} - {z}";
    }
    catch (Exception theException )
    {
        LoggingUtilities.encrich( theException, __FILE__, __LINE__, "someFunction", y, z );
   }
}

```
so when something went wrong - you got an email immediately saying "something went wrong" - with a stack trace - with the value of y and z - and everything up...   so by the time the trader gets to your desk, you don't have to turn around and go "duuuuuh - I don't know" - you can say - oh yeah - you put in a twap order 1231232, and we never expected the time to be an less than the settle time - so when it went minus, this thing overflowed.   We'll put a change in tonight - so it'll never happen again"

**Reply by En3Tho** [reply]
I see this as a runtime feature too. But while it seems totally plausible for you, this runtime toggle can actually break code because there is literally anything out there in the wild.

For example what if parameters have cyclic dependencies inside them and ToString'ing causes stack overflow?

What if parameters are represented as really big strings and exceptions are thrown quite often? This might lead to OutOfMemoryCrashes or logging bloat.

What if parameters have default ToString behaviour and basically give you nothing useful but you still pay for this feature?

Also you have to account for tremendous perf hits when wrapping every function in try catch blocks like those.

**Reply by darrenoakey** [reply]
you are correct - but 

- you only take those risks when an exception has already happened - so you aren't really "further" destabilising the app much in reality
- this would be a feature you can turn on - and I know our code base is very unlikely to have that sort of thing, because we have tests of everything, including every toString.. it is in fact guaranteed by our build
- to say a default toString might not get you anything...  well without this feature you get nothing guaranteed
- our world is etrading so performance is a key concern - again tests in the build gate guarantee certain latencies for certain functions etc... but we've been doing this for years - and are constantly profiling the code - the overhead of the try catch has not yet come up.  we often do get performance problems related to logging - but normally it's things like double -> string conversion and the like  [and we try to make sure that logging _work_ happens asynchronously].      while there clearly IS some cost in having a try catch in every function - I've never personally _felt_ it yet.
